test/integration/front_proxy_test.go

```diff
@@ -0,0 +1,316 @@
+	proxyGroupHeader := "X-Remote-Group"
+
+	certDir, err := ioutil.TempDir("", "frontproxycatestdir")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(certDir)
+	t.Logf("cert dir is %s\n", certDir)
+
+	frontProxyClientCA, err := createCA(certDir, frontProxyCAPrefix)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	proxyCert, err := createCert(proxyCertCommonName, certDir, frontProxyCAPrefix)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	masterConfig.AuthConfig.RequestHeader = &api.RequestHeaderAuthenticationOptions{
+		ClientCA:          frontProxyClientCA,
+		ClientCommonNames: []string{proxyCertCommonName},
+
+		// These don't get defaulted because we don't round trip config // TODO fix?
+		UsernameHeaders: []string{proxyUserHeader},
+		GroupHeaders:    []string{proxyGroupHeader},
+	}
+
+	clusterAdminKubeConfig, err := testserver.StartConfiguredMasterAPI(masterConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+	clusterAdminClientConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+	clusterAdminClient, err := testutil.GetClusterAdminClient(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	proxyHTTPHandler, err := newFrontProxyHandler(clusterAdminClientConfig.Host, masterConfig.ServingInfo.ClientCA, proxyUserHeader, proxyGroupHeader, proxyCert)
+	if err != nil {
+		t.Fatal(err)
+	}
+	proxyServer := httptest.NewServer(proxyHTTPHandler)
+	defer proxyServer.Close()
+	t.Logf("front proxy server is on %v\n", proxyServer.URL)
+
+	w, err := clusterAdminClient.Projects().Watch(kapi.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer w.Stop()
+
+	listProjectsRoleName := "list-projects-role"
+	if _, err := clusterAdminClient.ClusterRoles().Create(
+		&authorizationapi.ClusterRole{
+			ObjectMeta: kapi.ObjectMeta{Name: listProjectsRoleName},
+			Rules: []authorizationapi.PolicyRule{
+				authorizationapi.NewRule("list").Groups(projectapi.GroupName).Resources("projects").RuleOrDie(),
+			},
+		},
+	); err != nil {
+		t.Fatal(err)
+	}
+
+	for _, username := range []string{"david", "jordan"} {
+		projectName := username + "-project"
+		if _, err := testserver.CreateNewProject(clusterAdminClient, *clusterAdminClientConfig, projectName, username); err != nil {
+			t.Fatal(err)
+		}
+		waitForAdd(projectName, w, t)
+
+		// make it so that the user can list projects without any groups
+		if _, err := clusterAdminClient.ClusterRoleBindings().Create(
+			&authorizationapi.ClusterRoleBinding{
+				ObjectMeta: kapi.ObjectMeta{Name: username + "-clusterrolebinding"},
+				Subjects: []kapi.ObjectReference{
+					{Kind: authorizationapi.UserKind, Name: username},
+				},
+				RoleRef: kapi.ObjectReference{Name: listProjectsRoleName},
+			},
+		); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	for _, test := range []struct {
+		name             string
+		user             user.Info
+		isUnauthorized   bool
+		expectedProjects sets.String
+	}{
+		{
+			name:           "empty user",
```

Comment:
Hmm this may be the wrong behavior.  Should we expect the anonymous user or an unauthorized error when you do a `GET` to the front proxy but it does not set any headers on your behalf?  @liggitt